p2p/src/codec.rs

```diff
@@ -0,0 +1,268 @@
+use std::io;
+use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
+use tokio_util::codec::{Decoder, Encoder};
+
+pub struct MessageEncoder;
+
+impl Encoder<PeerMessage> for MessageEncoder {
+    type Error = io::Error;
+
+    fn encode(&mut self, item: PeerMessage, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        match item {
+            PeerMessage::Hello(u) => {
+                dst.put_u8(0); // Message type
+                const HELLO_MESSAGE_LEN: u64 = 2;
+                dst.put_u64(HELLO_MESSAGE_LEN);
+                dst.put_u16(u);
+            }
+            PeerMessage::Peers(p) => {
+                dst.put_u8(1); // Message type
+                let body = p.into_iter().fold(BytesMut::new(), |mut bytes, peer| {
+                    encode_peer_addr(peer, &mut bytes);
+                    bytes
+                });
+                dst.put_u64(body.len() as u64);
+                dst.put(body.as_ref());
+            }
+            PeerMessage::Data(data) => {
+                dst.put_u8(2); // Message type
+                dst.put_u64(data.len() as u64);
+                dst.put(data.as_slice());
+            }
+        }
+        Ok(())
+    }
+}
+
+fn encode_peer_addr(peer: PeerAddr, buf: &mut BytesMut) {
+    match peer.addr {
+        SocketAddr::V4(d) => {
+            buf.put_u8(4);
+            buf.put(&d.ip().octets()[..]);
+            buf.put_u16(d.port());
+        }
+        SocketAddr::V6(d) => {
+            buf.put_u8(6);
+            buf.put(&d.ip().octets()[..]);
+            buf.put_u16(d.port());
+            buf.put_u32(d.flowinfo());
+            buf.put_u32(d.scope_id())
+        }
+    }
+    buf.put(peer.id.0.as_bytes());
+}
+
+impl MessageEncoder {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+pub struct MessageDecoder {
+    state: DecodeState,
+}
+
+impl MessageDecoder {
+    pub fn new() -> Self {
+        MessageDecoder {
+            state: DecodeState::MessageType,
+        }
+    }
+}
+
+enum DecodeState {
+    MessageType,
+    Len(u8),
+    Body(u8, usize),
+}
+
+impl Decoder for MessageDecoder {
+    type Item = PeerMessage;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        match self.state {
+            DecodeState::MessageType => {
+                if src.is_empty() {
+                    return Ok(None);
+                }
+                let command_type = src.get_u8();
+                match command_type {
+                    0..=2 => {}
+                    _ => {
+                        return Err(io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            format!("Unknown command: {}", command_type),
+                        ))
+                    }
+                }
+                self.state = DecodeState::Len(command_type);
+                self.decode(src)
+            }
+            DecodeState::Len(m_type) => {
```

Comment:
~i don't see much value in cross-cutting of entities by factoring out a length prefix for multiple messages. E.g. for the Hello message there's now unnecessary length prefix because the message is of fixed-length. The only factored out piece we need is the tag for the message type, while each message type should bundle all the logic together. It may even make sense to have sub-Decoders for structuring this state machine cleaner.~

as i was writing the above thoughts, i began to see the value in having length prefix for the entire body, so we can simply wait once till all the bytes we need arrive, so we don't have to track all the intermediate states (which would be much nicer in a async/await function, but we'd need async fn in trait, which forces boxing of futures, so... nope).